1-js/03-code-quality/04-ninja-code/article.md

```diff
@@ -189,52 +187,57 @@ function render() {
   ...
   ...many lines...
   ...
-  ... // <-- a programmer wants to work with user here and...
+  ... // <-- یک برنامه‌نویس می‌خواهد در این‌جا با user کار کند و ...
   ...
 }
 ```
 
-A programmer who jumps inside the `render` will probably fail to notice that there's a local `user` shadowing the outer one.
+برنامه‌نویسی که تابع `render` را می‌خواند احتمالا متوجه نخواهد شد که یک `user` محلی وجود دارد که بر روی متغیر بیرونی سایه انداخته است.
 
-Then they'll try to work with `user` assuming that it's the external variable, the result of `authenticateUser()`... The trap is sprung! Hello, debugger...
+با فرض این که `user` یک متغیر خارجی است کار را ادامه می‌دهند. خواهند پنداشت این متغیر، نتیجه‌ی تابع `authenticateUser()` است. تله پهن شده است. درود بر اشکال‌زدایی.
 
 
-## Side-effects everywhere!
+## اثرات جانبی در همه‌جا!
 
-There are functions that look like they don't change anything. Like `isReady()`, `checkPermission()`, `findTags()`... They are assumed to carry out calculations, find and return the data, without changing anything outside of them. In other words, without "side-effects".
+توابعی وجود دارند که در ظاهر، چیزی را تغییر نمی‌دهند. مانند `isReady()`، `checkPermission()`، `findTags()` و ...
+فرض بر این است که این توابع، محاسباتی را انجام داده، داده‌ای را پیدا کرده و برمی‌گردانند، بدون این که چیزی را در خارج از حوزه‌ی خود تغییر دهند. به عبارت دیگر، این توابع بدون اثرات جانبی هستند.
 
-**A really beautiful trick is to add a "useful" action to them, besides the main task.**
+**یک ترفند واقعا زیبا این است که در کنار وظیفه‌ی اصلی توابع، یک عمل مفید دیگر را نیز به آن‌ها اضافه کنیم،**
 
-An expression of dazed surprise on the face of your colleague when they see a function named `is..`, `check..` or `find...` changing something -- will definitely broaden your boundaries of reason.
+حالت تعجب روی چهره‌ی یارانتان وقتی تابعی را می‌بینند که با `is..`، `check..` یا `find...` نام‌گذاری شده است
+و چیزی را هم تغییر می‌دهد، قطعا مرزهای منطق و استدلال شما را پهناورتر خواهد کرد.
 
-**Another way to surprise is to return a non-standard result.**
+**یک راه دیگر برای شگفت‌زده کردن، برگرداندن یک نتیجه‌ی غیراستاندارد است.**
 
-Show your original thinking! Let the call of `checkPermission` return not `true/false`, but a complex object with the results of the check.
+تفکر اصل خود را نشان دهید! بگذارید فراخوانی `checkPermission` مقدار `true/false` را برنگرداند، در عوض یک شی پیچیده با نتایچ بررسی را برگرداند.
 
-Those developers who try to write `if (checkPermission(..))`, will wonder why it doesn't work. Tell them: "Read the docs!". And give this article.
+توسعه‌دهندگانی که سعی کنند از `if (checkPermission(..))` استفاده کنند، حیرت‌زده خواهند شد که چرا این تابع به درستی عمل نمی‌کند. به آن‌ها بگویید: «مستندات را بخوانید!». و این مقاله را به آن‌ها معرفی کنید.
 
 
-## Powerful functions!
+## توابع قدرتمند
```

Comment:
Add `!` to match the English version
sorry if I'm being picky.

Reply:
Not at all :) That's a good thing.